rust/src/main.rs

```diff
@@ -0,0 +1,339 @@
+#![feature(test)]
+#![deny(unsafe_code)]
+
+extern crate itertools;
+extern crate mersenne_twister;
+extern crate num;
+extern crate rand;
+extern crate test;
+
+// Use BLAS directly
+#[cfg(feature = "direct_blas")]
+extern crate cblas;
+
+#[cfg(feature = "direct_blas")]
+extern crate blas_src;
+
+// Use ndarray (with BLAS implementation)
+#[cfg(not(feature = "direct_blas"))]
+#[macro_use(s)]
+extern crate ndarray;
+
+use std::time::{Duration, Instant};
+use std::u32;
+use std::fs::OpenOptions;
+use std::io::{BufWriter, Write};
+
+use test::black_box;
+use num::complex::Complex64;
+use rand::Rng;
+
+mod util;
+use util::{gen_rng, myrand};
+
+#[cfg(feature = "direct_blas")]
+mod direct_blas;
+#[cfg(feature = "direct_blas")]
+use direct_blas::{randmatstat, randmatmul, check_randmatmul};
+
+#[cfg(not(feature = "direct_blas"))]
+use ndarray::Array2;
+#[cfg(not(feature = "direct_blas"))]
+use util::fill_rand;
+#[cfg(not(feature = "direct_blas"))]
+use num::Zero;
+
+const NITER: u32 = 5;
+
+#[cfg(not(feature = "direct_blas"))]
+fn nrand<R: Rng>(shape: (usize, usize), rng: &mut R) -> Array2<f64> {
+    let mut m = Array2::zeros(shape);
+    fill_rand(&mut m, rng);
+    m
+}
+
+fn fib(n: i32) -> i32 {
+    let n = black_box(n); // prevent over-optimization
+    if n < 2 {
+        n
+    } else {
+        fib(n - 1) + fib(n - 2)
+    }
+}
+
+fn mandel(z: Complex64) -> u32 {
+    use std::iter;
+
+    iter::repeat(z)
+        .scan(z, |z, c| {
+            let current = *z;
+            *z = current * current + c;
+            Some(current)
+        })
+        .take(80)
+        .take_while(|z| z.norm() <= 2.0)
+        .count() as u32
+}
+
+fn mandelperf() -> Vec<u32> {
+    let mut m: Vec<u32> = vec![0; 21 * 26];
+
+    for ((i, j), mut v) in (0..21)
+        .flat_map(|i| (0..26).map(move |j| (i, j)))
+        .zip(&mut m)
+    {
+        *v = mandel(Complex64::new((j - 20) as f64 / 10.,
+            (i - 10) as f64 / 10.));
+    }
+
+    m
+}
+
+fn pisum() -> f64 {
+    let mut sum = 0.;
+    for _ in 0..500 {
+        sum = (1..10001)
+            .map(|k| {
+                let k = k as f64;
+                1. / (k * k)
+            })
+            .sum();
```

Comment:
This code doesn't seem reasonable to me.  Wouldn't it be faster and simpler to just write a loop to do the summation in Rust (since it is a compiled statically typed language, ordinary loops should be fast), rather than allocating a temporary array of summands and then calling `.sum()`?

Reply:
There is no allocation there: the abstractions in Rust have zero cost. This is the idiomatic way to write this code in Rust.

Reply:
Thanks for the clarification; I didn't realize that `map` was lazy in Rust.